autonat/autonat-v2.md

```diff
@@ -0,0 +1,137 @@
+# AutonatV2: spec
+
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+|-----------------|--------------------------|--------|-----------------|
+| 1A              | Working Draft            | Active | r0, 2023-04-11  |
+
+Authors: [@sukunrt], [@marten-seemann]
+
+Interest Group: [@marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@sukunrt]: https://github.com/sukunrt
+
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT / firewall or if it is
+publicly reachable. Knowing its NAT status is essential for the node to be
+well-behaved in the network: A node that's behind a doesn't need to
+advertise its (undialable) addresses to the rest of the network, preventing
+superfluous dials from other peers. Furthermore, it might actively seek to
+improve its connectivity by finding a relay server, which would allow other
+peers to establish a relayed connection.
+
+`autonat v1` protocol was designed for a node to determine whether the node
+was publically rechable. However, it could not be tweaked to obtain more 
+granular information about which of the nodes addresses were publically 
+reachable. `autonat v2` allows nodes to determine reachability for individual
+addresses. Using `autonat v2` nodes can build an address pipeline where they
+can test individual addresses discovered by different sources like identify, 
+upnp mappings, circuit addresses etc for reachability.
+
+
+## AutoNAT V2 Protocol
+
+The AutoNAT V2 Protocol uses the Protocol ID `/libp2p/autonat/2.0.0`. The node
+wishing to determine reachability of a given address sends a `DialRequest` 
+message. The `DialRequest` message contains a single multiaddress and a nonce 
+of 8 random bytes. Upon receiving  this message, the peer dials this address 
```

Comment:
If we specify that it's 8 bytes, should we just make a uint64?

Reply:
I'm fine either way, do you think uint64 would be easier to handle?

Reply:
Probably, if we use 8 bytes. However, I'm wondering if we might have any use case for longer sequences, and should leave the number of bytes unspecified (or just recommend a minimum).

Reply:
I've changed this to uint64 because that does seem easier to handle for implementations. I'm not sure how to decide here, I can't think of a need for longer sequences, but that will be more flexible. Keeping this open for a bit

Reply:
Yes. Curious to hear what the other reviewers think.